File: parte3/gauss-com-pivotamento-completo.m
% Metodo de Gauss com pivotamento completo
% Nesse metodo serao efetuadas trocas de linhas, caso necessario

% clear
% clc
format short

% a = [1 -1 1; 2 3 -1; -3 1 1];
% b = [1 4 -1];
a = [1 1 0 1; 2 1 -1 -1; -1 -2 3 -1; 3 -1 -1 2];
b = [2 1 4 -3];
n = length(b)
o = [1 2 3 4]

for k=1:(n - 1)
    % Verificar o maior elemento em modulo da coluna
    maior = abs(a(o(k),k));
    p = k; % Linha do pivo
    for i=(k + 1):n
        if (abs(a(o(i),k)) > maior)
            maior = abs(a(o(i),n));
            p = i; % Linha do novo pivo (O maior em modulo)
        end
    end

    % Trocar linha k por linha maior no vetor o
    if (p > k)
        temp = o(k);
        o(k) = o(p);
        o(p) = temp;
    end


    % Gauss, zerando elementos abaixo do pivo
    % Triangularizacao
    for i=(k + 1):n

        mult = a(o(i),k) / a(o(k),k);

        for j=k:n

            a(o(i),j) = a(o(i),j) - mult * a(o(k),j);

        end

        b(o(i)) = b(o(i)) - mult * b(o(k));

    end

end
a
b
ca = a; % Para calcular o residuo
cb = b; % Para calcular o residuo
o % Vetor de ordenamento das trocas de linhas

% Retrosubstituicao

x(n) = b(o(n)) / a(o(n),n);

for i=(n - 1):-1:1

  soma = 0;

  for j=(i + 1):n

    soma = soma + x(j)*a(o(i),j);

  end

  x(i) = (b(o(i)) - soma) / a(o(i),i);

end

x

% Residuo -- Nao eh comum calcular o residuo em metodos diretos, como esse, pois
% eh feita a copia do sistema original em memoria

r = abs(cb' - ca * x') % x e b como vetor coluna apostrofo faz o transposto

File: parte3/gauss-com-pivotamento-completo.py

# As linhas e colunas serao representadas com os valores
# de 0 ate n. Entao em um exemplo de matriz 3x3, suas linhas
# e colunas serao representadas de 0 ate 2

def gauss_com_pivoteamento(a, b, n):
    # Recebe a dimensão da matriz
    # i = k + 1
    # j = k



    # Escalonamento da Matriz
    for k in range(n): # Varia de 1:n-1 
        # Verificar se a linha k tem o maior coeficiente na coluna k

        # Pega o pivo
        pivo = abs(a[k][k])
        p=k
        for i in range(k+1, n): # Varia k+1:n
            if (abs(a[i][k]) > pivo):
                pivo = a[i][k]
                p = i

        print("Pivo:", pivo)
        print("P:", p)
        # Cria um vetor "o" que contem a ordem das linhas. 
        # Depois todas as referencias das linhas serao em 
        # funcao do vetor "o", substituindo i e k por o[i] e o[k]


        print("\n\n")

        for i in range(k+1, n+1): # Varia de k+1:n

            mult = a[i-1][k-1] / a[k-1][k-1]

            for j in range(k, n+1): # Varia de k:n
                # aij = aij - (aik / akk)*akj
                a[i-1][j-1] = a[i-1][j-1] - mult * a[k-1][j-1]

        # bi = bi - (aik / akk)*bkbirge_vieta
        b[i-1] = b[i-1] - mult * b[k-1]

    print(a)
    print(b)

    # Retrosubstituicao para resolver o sistema linear
    x = [0] * n
    x[n-1] = b[n-1] / a[n-1][n-1]


    for i in range(n, 0, -1):
        soma = 0

        for j in range(i+1, n):
            soma = soma + a[i-1][j-1] * x[j-1]

        x[i-1] = (b[i-1] - soma) / a[i-1][i-1]

    print(x)




if __name__ == "__main__":
    # Define a list of lists
    a = [
        [1, -1, 1],
        [2, 3, -1],
        [-3, 1, 1]
    ]
    b = [1, 4, -1]
    gauss_com_pivoteamento(a, b, len(a))